src/validation-engine.ts

```diff
@@ -0,0 +1,238 @@
+    return this.buildValidationResults(allValidations);
+  }
+
+  validateField(
+    values: any,
+    key: string,
+    value: any
+  ): Promise<ValidationResult> {
+    this.asyncValidationInProgressCount++;
+    const asyncValidationPromise = this.fireFieldValidations(
+      values,
+      key,
+      value
+    );
+
+    // TODO: de we need to add some additional check in the promise
+    // e.g. fail or something like that?
+    asyncValidationPromise.then(() => this.asyncValidationInProgressCount--);
+
+    return asyncValidationPromise;
+  }
+
+  private fireFieldAndRecordValidations = (
+    values: any
+  ): Promise<ValidationResult>[] => {
+    let fieldValidationResults: Promise<
+      ValidationResult
+    >[] = fireAllFieldsValidations(
+      values,
+      Object.keys(this.validationsPerField),
+      this.fireFieldValidations
+    );
+
+    // Now global form validations
+    if (this.validationsGlobalForm.length > 0) {
+      fieldValidationResults = [
+        ...fieldValidationResults,
+        ...this.validateGlobalFormValidations(values),
+      ];
+    }
+    return fieldValidationResults;
+  };
+
+  private buildValidationResults = (
+    validations: Promise<ValidationResult>[]
+  ): Promise<FormValidationResult> => {
+    this.asyncValidationInProgressCount++;
+    return new Promise<FormValidationResult>((resolve, reject) => {
+      // Once all the single field validations have been resolved
+      // resolve the fullFormValidatePromise
+      // TODO: Finally issue race condition unit tests
+      Promise.all(validations)
+        .then((fieldValidationResults: ValidationResult[]) => {
+          this.asyncValidationInProgressCount--;
+          const formValidationResult = buildFormValidationResult(
+            fieldValidationResults
+          );
+          resolve(formValidationResult);
+        })
+        .catch(result => {
+          this.asyncValidationInProgressCount--;
+          // Build failed validation Result
+          const errorInformation = `Uncontrolled error when validating full form, check custom validations code`;
+          console.log(errorInformation);
+          reject(result);
+        });
+    });
+  };
+
+  public fireFieldValidations = (
+    values: any,
+    key: string,
+    value: any
+  ): Promise<ValidationResult> => {
+    this.asyncValidationInProgressCount++;
+
+    const fieldValidationResultPromise = new Promise<ValidationResult>(
+      (resolve, reject) => {
+        if (!this.isFieldKeyMappingDefined(key)) {
+          this.asyncValidationInProgressCount--;
+          resolve(createDefaultValidationResult());
+        } else {
+          fireSingleFieldValidations(
+            values,
+            value,
+            this.validationsPerField[key]
+          )
+            .then((fieldValidationResult: ValidationResult) => {
+              this.asyncValidationInProgressCount--;
+              if (fieldValidationResult) {
+                fieldValidationResult.key = key;
```

Comment:
I think we should not provide the `key` property to result. `key` is an internal property for us, but when we use the library a got the result we have something like:

```
const onUpdateField = (fieldName, value) => {
    setLogin({
      ...login,
      [fieldName]: value,
    });

    validation
      .validateField(login, fieldName, value)
      .then(validationResult => {
        setLoginErrors({
          ...loginErrors,
          [fieldName]: validationResult 
        });
      });
  };
```
We always have the `fieldName` property available since we are executing `validateField` with it.

Reply:
Can you elaborate a proposal on this?

Maybe keep two interface internalValidationResult and ValidationResult?

Do you think it could be a good idea to keep this as a separate case and open a separate branch for this?